include/tsar/Analysis/Memory/MemorySetInfo.h

```diff
@@ -157,6 +192,93 @@ template<> struct MemorySetInfo<MemoryLocationRange> {
+      return false;
+    if (LHS.DimList.size() == 0) {
+      return sizecmp(getUpperBound(LHS), getLowerBound(RHS)) >= 0 &&
+        sizecmp(getLowerBound(LHS), getUpperBound(RHS)) <= 0;
+    }
+    bool Result = true;
+    std::size_t JoinableDimCount = 0;
+    for (std::size_t I = 0; I < LHS.DimList.size(); ++I) {
+      auto &Left = LHS.DimList[I];
+      auto &Right = RHS.DimList[I];
+      assert(Left.TripCount > 0 && Right.TripCount > 0 &&
+          "Trip count of dimension must be positive!");
+      if (Left == Right)
+        continue;
+      auto LeftEnd = Left.Start + Left.Step * (Left.TripCount - 1);
+      auto RightEnd = Right.Start + Right.Step * (Right.TripCount - 1);
+      if (Left.Step != Right.Step ||
+          (Left.Start - Right.Start) % Left.Step != 0 ||
+          (LeftEnd < Right.Start && (Right.Start - LeftEnd) != Left.Step) ||
+          (RightEnd < Left.Start && (Left.Start - RightEnd) != Left.Step)) {
+        Result = false;
+        break;
+      }
+      ++JoinableDimCount;
+    }
+    return Result & (JoinableDimCount <= 1);
+  }
+  static inline bool join(MemoryLocationRange &LHS,
+      const MemoryLocationRange &RHS) {
+    if (LHS.DimList.size() != LHS.DimList.size())
+      return false;
+    if (LHS.DimList.size() == 0) {
+      if (sizecmp(getUpperBound(LHS), getUpperBound(RHS)) < 0) {
+        setUpperBound(getUpperBound(RHS), LHS);
+        return true;
+      }
+      if (sizecmp(getLowerBound(LHS), getLowerBound(RHS)) > 0) {
+        setLowerBound(getLowerBound(RHS), LHS);
+        return true;
+      }
+      return false;
+    }
+    bool isChanged = false;
+    for (size_t I = 0; I < LHS.DimList.size(); I++) {
+      auto &To = LHS.DimList[I];
+      auto &From = RHS.DimList[I];
+      if (From.Start < To.Start) {
+        To.TripCount += (To.Start - From.Start) / From.Step;
+        To.Start = From.Start;
+        isChanged = true;
+      }
+      auto ToEnd = To.Start + To.Step * (To.TripCount - 1);
+      auto FromEnd = From.Start + From.Step * (From.TripCount - 1);
+      if (FromEnd > ToEnd) {
+        To.TripCount += (FromEnd - ToEnd) / From.Step;
+        isChanged = true;
+      }
+    }
+    return isChanged;
+  }
+  static inline bool hasIntersection(const MemoryLocationRange &LHS,
+      const MemoryLocationRange &RHS) {
+    if (LHS.DimList.size() != RHS.DimList.size())
+      return false;
+    if (LHS.DimList.size() == 0 && RHS.DimList.size() == 0) {
+      return sizecmp(getUpperBound(LHS), getLowerBound(RHS)) > 0 &&
+             sizecmp(getLowerBound(LHS), getUpperBound(RHS)) < 0;
+    }
+    MemoryLocationRange Int;
+    return MemoryLocationRangeEquation::intersect(LHS, RHS, Int,
+        nullptr, nullptr);
+  }
+  static inline bool intersect(
+      const MemoryLocationRange &LHS,
+      const MemoryLocationRange &RHS,
+      MemoryLocationRange &I,
+      llvm::SmallVector<MemoryLocationRange, 0> *L,
+      llvm::SmallVector<MemoryLocationRange, 0> *R) {

```

Comment:
```suggestion
      llvm::SmallVectorImpl<MemoryLocationRange> *L,
      llvm::SmallVectorImpl<MemoryLocationRange> *R) {
```

Может еще значения по умолчанию добавить `nullptr`, чтобы не писать явно, когда нужно только пересечение?